src/operator/contrib/two_bit_quantize-inl.h

```diff
@@ -0,0 +1,340 @@
+#define MXNET_OPERATOR_CONTRIB_TWO_BIT_QUANTIZE_INL_H_
+
+#include <mxnet/operator_util.h>
+#include <vector>
+#include <limits>
+#include "../elemwise_op_common.h"
+#include "../mshadow_op.h"
+#include "../mxnet_op.h"
+
+namespace mxnet {
+namespace op {
+
+struct init_mem_2bit {
+  // Initialize output array
+  MSHADOW_XINLINE static void Map(int i, float* out) {
+    *(out+i) = 0;
+  }
+};
+
+struct TwoBitParam : public dmlc::Parameter<TwoBitParam> {
+  float pos_threshold, neg_threshold;
+  DMLC_DECLARE_PARAMETER(TwoBitParam) {
+    DMLC_DECLARE_FIELD(neg_threshold)
+      .set_default(-0.1)
+      .describe("Threshold to quantize negative values. "
+                  "Has to be less than 0");
+    DMLC_DECLARE_FIELD(pos_threshold)
+      .set_default(0.1)
+      .describe("Threshold to quantize positive values. "
+                  "Has to be greater than 0");
+  }
+};
+
+template<typename xpu>
+void Create2BitArrayCompute(const nnvm::NodeAttrs& attrs,
+                            const OpContext& ctx,
+                            const std::vector<TBlob>& inputs,
+                            const std::vector<OpReqType>& req,
+                            const std::vector<TBlob>& outputs) {
+  // For now, this method can only compress the float data
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  // Init the memory of output to 0x00000000
+  mxnet_op::Kernel<init_mem_2bit, xpu>::Launch(s, outputs[0].Size(),
+                              outputs[0].dptr<float>());  // compressed array
+}
+
+inline bool Create2BitArrayShape(const nnvm::NodeAttrs& attrs,
+                                 std::vector<TShape> *in_attrs,
+                                 std::vector<TShape> *out_attrs) {
+  // 0. input array
+  CHECK_EQ(in_attrs->size(), 1U);
+  // 0. output array
+  CHECK_EQ(out_attrs->size(), 1U);
+  // check input
+  CHECK(!shape_is_none(in_attrs->at(0)));
+  // output
+  int shape = in_attrs->at(0).Size() % 16 == 0 ?
+                    in_attrs->at(0).Size() / 16 + 3:
+                    in_attrs->at(0).Size() / 16 + 4;
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape{shape});
+  return true;
+}
+
+inline bool Create2BitArrayType(const nnvm::NodeAttrs &attrs,
+                                std::vector<int> *in_attrs,
+                                std::vector<int> *out_attrs) {
+  // 0. input array
+  CHECK_EQ(in_attrs->size(), 1U);
+  // 0. output array
+  CHECK_EQ(out_attrs->size(), 1U);
+  // check input
+  CHECK_EQ((*in_attrs)[0], mshadow::kFloat32)
+    << "`create_2bit_` only supports float32 input for now";
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::kFloat32);
+  return true;
+}
+
+struct init_threshold_2bit {
+  MSHADOW_XINLINE static void Map(int i,
+                                  float *out,
+                                  const float neg_threshold,
+                                  const float pos_threshold,
+                                  int size) {
+    // The first two elements in output are thresholds
```

Comment:
This seem strange -- i is not used at all?  Can you explain?

Reply:
I only needed to launch 1 kernel here for initializing these meta variables. i can only be 1, we don't need it. Doing this outside this MSHADOW_XINLINE means we can't access data on GPUs right? Is there a better way?